### PR TITLE
Grouped convolution

### PR DESCRIPTION
**Pull request by anaruse**
The aim of this PR is to enable use of grouped convolution from Chainer.

An argument `group` is added to `Convolution2D` class, so you can easily use grouped convolution by just adding the argument `group` with group count to convolution layers in your network like below.

    conv1 = L.Convolution2D(..., group=8)

This variant of convolution works like a depthwise convolution when group count is equal to number of input/output channels.

You can use this convolution on GPU with cuDNN7, GPU with cupy and CPU with numpy. Please note that https://github.com/cupy/cupy/pull/581 is necessary to use this convolution with cuDNN7. 

AIs
* Add tests

### CODE REVIEW

**Comment by anaruse** [comment]
Tests for grouped convolution and ResNeXt example using grouped convolution were added.

**Comment by anaruse** [comment]
Resolved conflicts with master branch.

File: chainer/functions/connection/convolution_2d.py [hunk 1]
```diff
@@ -440,9 +559,10 @@ def convolution_2d(x, W, b=None, stride=1, pad=0, cover_all=False, **kwargs):
         "supported anymore. "
         "Use chainer.using_config('cudnn_deterministic', value) "
         "context where value is either `True` or `False`.")
-    dilate, = argument.parse_kwargs(kwargs, ('dilate', 1))
+    dilate, group = argument.parse_kwargs(kwargs, ('dilate', 1), ('group', 1))
```
Comment (takagi) [inline chainer/functions/connection/convolution_2d.py]: ditto

File: chainer/functions/connection/deconvolution_2d.py [hunk 2]
```diff
@@ -42,12 +42,14 @@ def __init__(self, stride=1, pad=0, outsize=None, **kwargs):
             "the gradient w.r.t. x is automatically decided during "
             "backpropagation."
         )
-        dilate, = argument.parse_kwargs(kwargs, ('dilate', 1))
+        dilate, group = argument.parse_kwargs(kwargs,
+                                              ('dilate', 1), ('group', 1))
```
Comment (takagi) [inline chainer/functions/connection/deconvolution_2d.py]: ditto

File: chainer/functions/connection/deconvolution_2d.py [hunk 3]
```diff
@@ -366,9 +427,10 @@ def deconvolution_2d(x, W, b=None, stride=1, pad=0, outsize=None, **kwargs):
         "supported anymore. "
         "Use chainer.using_config('cudnn_deterministic', value) "
         "context where value is either `True` or `False`.")
-    dilate, = argument.parse_kwargs(kwargs, ('dilate', 1))
+    dilate, group = argument.parse_kwargs(kwargs, ('dilate', 1), ('group', 1))
```
Comment (takagi) [inline chainer/functions/connection/deconvolution_2d.py]: ditto

File: chainer/links/connection/convolution_2d.py [hunk 4]
```diff
@@ -108,7 +108,8 @@ def __init__(self, in_channels, out_channels, ksize=None, stride=1, pad=0,
             "supported anymore. "
             "Use chainer.using_config('cudnn_deterministic', value) "
             "context where value is either `True` or `False`.")
-        dilate, = argument.parse_kwargs(kwargs, ('dilate', 1))
+        dilate, group = argument.parse_kwargs(kwargs,
+                                              ('dilate', 1), ('group', 1))
```
Comment (takagi) [inline chainer/links/connection/convolution_2d.py]: ditto

**Comment by takagi** [comment]
Sorry for leaving this PR for long. Would you check the comments?

File: chainer/functions/connection/convolution_2d.py [hunk 5]
```diff
@@ -39,12 +39,14 @@ def __init__(self, stride=1, pad=0, cover_all=False, **kwargs):
             "the gradient w.r.t. x is automatically decided during "
             "backpropagation."
         )
-        dilate, = argument.parse_kwargs(kwargs, ('dilate', 1))
+        dilate, group = argument.parse_kwargs(kwargs,
+                                              ('dilate', 1), ('group', 1))
```
[the remark this answers is not in the document]
Comment (anaruse) [reply chainer/functions/connection/convolution_2d.py]: Thank you for your comment.

I also think that `group` argument has sufficient generality to be treated as standard one. will move it from `parse_kwargs` to `__init__`.

File: chainer/functions/connection/convolution_2d.py [hunk 6]
```diff
@@ -168,19 +185,61 @@ def forward_gpu(self, inputs):
...
+        xp = cuda.get_array_module(x)
+
+        _x = x.reshape(N, G, iCg, iH, iW)
+        _x = xp.rollaxis(_x, 1)  # (G, N, iCg, iH, iW)
+        _W = W.reshape(G, oCg, iCg, kH, kW)
+        if b is not None:
+            _b = b.reshape(G, oCg)
+
+        _ys = []
+        for g in range(G):
```
[the remark this answers is not in the document]
Comment (anaruse) [reply chainer/functions/connection/convolution_2d.py]: Thanks, will use `six.moves.rage` here.

File: chainer/functions/connection/convolution_2d.py [hunk 7]
```diff
@@ -168,19 +185,61 @@ def forward_gpu(self, inputs):
...
+            _bg = None if b is None else _b[g, ]
+            if xp is numpy:
+                _y = self._forward_cpu_core(_x[g, ], _W[g, ], _bg)
+            else:
+                _y = self._forward_gpu_core(_x[g, ], _W[g, ], _bg)
+            _ys.append(_y)
+
+        y = xp.stack(_ys, axis=1)  # (N, G, oCg, oH, oW)
+        _, _, _, oH, oW = y.shape
+        y = y.reshape((N, oC, oH, oW))
```
[the remark this answers is not in the document]
Comment (anaruse) [reply chainer/functions/connection/convolution_2d.py]: Thanks, will fix in that way.

**Comment by anaruse** [comment]
Thank you for your comments, @takagi .

I fixed the code in the suggested way. The conflicts with master branch were also resolved.
Please note that you should apply https://github.com/cupy/cupy/pull/785 prior to testing this PR on GPU.

**Comment by takagi** [comment]
jenkins, test this please.

**Comment by takagi** [comment]
Our CI environment performs multi-environment tests including older cuDNN versions, causing failures in `TestConvolution2DCudnnCall` and `TestDeconvolution2DCudnnCall`.

Would you add the condition if cuDNN should be used or not?  Around [test_convolution_2d.py#L298](https://github.com/anaruse/chainer/blob/b7d9c7baed81fddfe3fba1f14521b407fffe996e/tests/chainer_tests/functions_tests/connection_tests/test_convolution_2d.py#L298) and [test_deconvolution_2d.py#L259](https://github.com/anaruse/chainer/blob/b7d9c7baed81fddfe3fba1f14521b407fffe996e/tests/chainer_tests/functions_tests/connection_tests/test_deconvolution_2d.py#L259) would be good for that.

**Comment by anaruse** [comment]
Thank you for your testing, @takagi 

I've just updated the PR so that it passes the tests in `test_convolution_2d.py` and `test_deconvolution_2d.py` with cuDNN 6 or older. I've confirmed that it passes the tests with the following combinations:
* CUDA 9.0 and cuDNN 7.0
* CUDA 8.0 and cuDNN 6.0

Please test again.

**Comment by takagi** [comment]
jenkins, test this please.

**Comment by takagi** [comment]
Thanks, the CI passes and it now shows green lamp. But I'm afraid that I need to ask you to resolve the conflict again. The change that brings the conflict cleans the cpu/gpu forward/backward paths, so would you rearrange your changes to follow the way?

**Comment by anaruse** [comment]
All right, Chainer is really fast changing target, isn't it?
Anyway, I've resolved the conflicts to the latest master. Please check again.

**Comment by takagi** [comment]
Thanks! Yeah, the conflicting change is especially interesting because it is preparation for Intel iDeep support. Please let me check the change.

**Comment by takagi** [comment]
LGTM!

**Comment by takagi** [comment]
The backport follows #3848.

**Pull request merged**